packages/plugins/typed-variable-modal/src/Modal.js

```diff
@@ -0,0 +1,378 @@
+    if (this.shouldCloseOnOverlayClick) {
+      this.hideEvent = this.addEvent_(widgetDiv, 'click', this, this.hide);
+      this.modalEvent = this.addEvent_(this.htmlDiv_, 'click', this, (e) => {
+        e.stopPropagation();
+      });
+    }
+    Blockly.utils.dom.addClass(this.htmlDiv_, 'blockly-modal-open');
+    widgetDiv.appendChild(this.htmlDiv_);
+  }
+
+  /**
+   * Disposes of any events or dom-references belonging to the editor.
+   * @protected
+   */
+  widgetDispose_() {
+    const widgetDiv = Blockly.WidgetDiv.DIV;
+    widgetDiv.style.width = 'auto';
+    widgetDiv.style.height = 'auto';
+    if (this.shouldCloseOnOverlayClick) {
+      Blockly.unbindEvent_(this.hideEvent);
+      Blockly.unbindEvent_(this.modalEvent);
+    }
+    Blockly.utils.dom.removeClass(this.htmlDiv_, 'blockly-modal-open');
+  }
+
+  /**
+   * Handle when the user goes to the previous focusable element.
+   * @param {KeyboardEvent} e The keydown event.
+   * @private
+   */
+  handleBackwardTab_(e) {
+    if (document.activeElement === this.firstFocusableEl_) {
+      e.preventDefault();
+      this.lastFocusableEl_.focus();
+    }
+  }
+
+  /**
+   * Handle when the user goes to the next focusable element.
+   * @param {KeyboardEvent} e The keydown event.
+   * @private
+   */
+  handleForwardTab_(e) {
+    if (document.activeElement === this.lastFocusableEl_) {
+      e.preventDefault();
+      this.firstFocusableEl_.focus();
+    }
+  }
+
+  /**
+   * Handles keydown event for a Blockly modal. Handles forward tab, backward
+   * tab, and escape button.
+   * @param {KeyboardEvent} e The keydown event.
+   * @private
+   */
+  handleKeyDown_(e) {
+    if (e.keyCode === Blockly.utils.KeyCodes.TAB) {
+      // If there are no elements or there is one element don't wrap.
+      if (!this.firstFocusableEl_ ||
+          this.firstFocusableEl_ === this.lastFocusableEl_) {
+        e.preventDefault();
+        e.stopPropagation();
+        return;
+      }
+      if (e.shiftKey) {
+        this.handleBackwardTab_(e);
+      } else {
+        this.handleForwardTab_(e);
+      }
+    } else if (e.keyCode === Blockly.utils.KeyCodes.ESC
+        && this.shouldCloseOnEsc) {
+      this.hide();
+    }
+    e.stopPropagation();
+  }
+
+  /**
+   * Helper method for adding an event.
+   * @param {!Element} node Node upon which to listen.
+   * @param {string} name Event name to listen to (e.g. 'mousedown').
+   * @param {Object} thisObject The value of 'this' in the function.
+   * @param {!Function} func Function to call when event is triggered.
+   * @protected
+   */
+  addEvent_(node, name, thisObject, func) {
+    const event = Blockly.bindEventWithChecks_(node, name, thisObject, func);
+    this.boundEvents_.push(event);
+  }
+
+  /**
+   * Create all the dom elements for the modal.
+   */
+  render() {
+    /*
+     * Creates the Modal. The generated modal looks like:
+     * <div class="blockly-modal-container" role="dialog">
```

Comment:
I know that we've split up the "content" of the modal into a header / content / footer, but these are just semantic divisions. 
The "content" of the modal (ie: the div that gets styled to be in the center) in this case is the container div, and i don't see a div for the overlay (in case you wanted to style that differently). 

Could we have an overlay div and a content div.
```
<div class="blocklyModalOverlay">
  <div class="blocklyModalContent">
      <header class="blocklyModalHeader"></header>
      <main class="blocklyModalMainContent"></main>
      <footer class="blocklyModalHeader"></footer>
  </div>
</div>
```
or similar.

This lets you style the modal like this: 
![Screen Shot 2020-04-01 at 2 43 50 PM](https://user-images.githubusercontent.com/16690124/78189726-71cc9180-7427-11ea-87a4-edaad1cfe618.png)

Reply:
I think the "overlay" is the widget div. When we show the modal we add it into the widget div. I could add a class to the widget div so they could do something similar to that?

Reply:
(In summary), there's no way to style the overlay.

Reply:
You could add a style to the widget div, but I think and I'm seeing this in the widgetDispose, that's adding a lot of extra complexity on widget cleanup.
Why not just add another div, say blocklyModalOverlay and set all event handlers, and styling on that.

Making the widget div just a container for housing things, that is simply cleared on hideChaff. The other benefit of that, is you wouldn't have to compete with .blocklyWidgetDiv CSS elsewhere in the editor for what you're trying to achieve.

Reply:
Ok updated.